docs/proposal/20240308-vpc-infra-creation.md

```diff
@@ -0,0 +1,393 @@
+	// +optional
+	PublicGateways []*VPCResourceReference `json:"publicGateways,omitempty"`
+
+	// securityGroups references the VPC Security Groups for the cluster.
+	// +optional
+	SecurityGroups []*VPCResourceReference `json:"securityGroups,omitempty"`
+
+	// vpc references the IBM Cloud VPC.
+	// +optional
+	VPC *VPCResourceReference `json:"vpc,omitempty"`
+}
+
+// VPCResourceReference identifies a resource by id and type.
+type VPCResourceReference struct {
+	// controllerCreated indicates whether the resource is created by the CAPI controller.
+	ControllerCreated bool `json:"controllerCreated,omitempty"`
+
+	// crn defines the IBM Cloud CRN of the resource.
+	// +optional
+	CRN *string `json"crn,omitempty"`
+
+	// id defines the ID of the IBM Cloud resource.
+	ID string `json:id"`
+
+	// type defines the type of IBM Cloud resource.
+	Type VPCResourceType `json:type"`
+}
+
+type VPCResourceType string
+
+const (
+	VPCResourceTypeCOS               VPCResourceType = VPCResourceType("cos")
+	VPCResourceTypePublicGateway     VPCResourceType = VPCResourceType("publicgateway")
+	VPCResourceTypeSecurityGroup     VPCResourceType = VPCResourceType("securitygroup")
+	VPCResourceTypeSecurityGroupRule VPCResourceType = VPCResourceType("securitygrouprule")
+	VPCResourceTypeSubnet            VPCResourceType = VPCResourceType("subnet")
+	VPCResourceTypeVPC               VPCResourceType = VPCResourceType("vpc")
+)
+```
+
+
+#### Share existing common resource definitions
+Rather than duplicate code and also make references to existing resources
+complex, we also propose relocating existing resource references into a
+common file. This prevents the duplication of code in PowerVS and VPC
+definitions, but also removes the direct dependence on PowerVS definitions,
+placing them into a common location (preventing unexpected dependencies).
+
+Relocate the following resource definitions in IBMPowerVSCluster
+```go
+// VPCResourceReference is a reference to a specific VPC resource by ID or Name
+// Only one of ID or Name may be specified. Specifying more than one will result in
+// a validation error.
+type VPCResourceReference struct {
+	// ID of resource
+	// +kubebuilder:validation:MinLength=1
+	// +optional
+	ID *string `json:"id,omitempty"`
+
+	// Name of resource
+	// +kubebuilder:validation:MinLength=1
+	// +optional
+	Name *string `json:"name,omitempty"`
+
+	// IBM Cloud VPC region
+	Region *string `json:"region,omitempty"`
+}
+
+// CosInstance represents IBM Cloud COS instance.
+type CosInstance struct {
+	// PresignedURLDuration defines the duration for which presigned URLs are valid.
+	//
+	// This is used to generate presigned URLs for S3 Bucket objects, which are used by
+	// control-plane and worker nodes to fetch bootstrap data.
+	//
+	// When enabled, the IAM instance profiles specified are not used.
+	// +optional
+	PresignedURLDuration *metav1.Duration `json:"presignedURLDuration,omitempty"`
```

Comment:
We had hard time creating a url with presignedduration, may be an experiment to verify it works and on cofirmation of it we can add this field.

Reply:
I think we tried the TOKEN based authorisation and it works fine, may be you can consider using that to keep it simple with the IAM token instead of access key/secret key for this presignedURL stuff.

Reply:
+1 we already have pushed code using IAM token same can be adopted here.

Reply:
This was simply suggesting moving the existing `CosInstance` struct out of `ibmpowervscluster_types.go` into a "shared" location, so VPC does not redefine the resource and so VPC is not referencing a `CosInstance` in PowerVS managed files.

If you wish to drop or make this field deprecated, I can do that. If you wish to have a new unique resource for COS, I can implement that in a shared location (that you can choose to migrate to, if you desire).